### ISSUE TITLE
Add Python 3.x support

### ISSUE THREAD

**Issue by valentjn** [SGpp/SGpp#53]
In GitLab by @valentjn on Nov 28, 2017, 15:30

This has long been on our wish list, but chances are that we are finally able to make the transition to Python 3.x. Yay! (I mean, Python 3.0 was released way back in 2008...)

SWIG supports Python 3.x for years now, and the only thing that prevented us to switch to Python 3.x was the infamous SCons, which only supported Python 2.x. Behold! SCons 3.0.0 was released in September 2017 and it's the first Python-3.x-compatible version. It's not yet in the Ubuntu package sources, but it will be in Bionic's (18.04 LTS).

For the SWIG side, I've already added an ugly SCons switch `USE_PYTHON3_FOR_PYSGPP` in d48a0e885 (with a fix in c5d28a586), but that's only a temporary solution. Probably more sophisticated would be to check the Python version SCons is using, and taking that as a default for SWIG. We could still support to override that with a SCons switch.

The alternative, brute-force approach would be for SG++ to require Python 3.x (and, thus, >= SCons 3.0) beginning with the next release.

`pysgpp` seems to work with Python 3.x. However, there are a lot of Python 2.x scripts in the `examples` and `python` subdirectories of the modules. Best would be to rewrite them so that either they run with Python 3.x or they run with Python 2.x and 3.x (`from __future__ import ...`). This could be the opportunity to prune some old scripts which don't work anymore anyway.

Ideas? Tagging @fabianfranzelin as he's using Python and `pysgpp` a lot :smile:

**Comment by valentjn** [comment]
In GitLab by @schobepr on Dec 11, 2017, 09:43

I tried to use scons 3 with USE_PYTHON3_FOR_PYSGPP=1 on windows using anaconda3 distribution (Py3.6 and Py3.7) to no avail:
```
$ ./schobepr_scons_build_mingw.bat

E:\MATLAB\SparseGridsLifecycle\BBSG\Examples\external\SGpp>scons -j 4   VERBOSE=0   SG_ALL=0   SG_BASE=1   SG_OPTIMIZATION=1   OPT=1   USE_EIGEN=1   USE_MATLAB=1   COMPILER=gnu   SG_PYTHON=1   USE_PYTHON3_FOR_PYSGPP=1   RUN_PYTHON_TEST=0   BOOST_INCLUDE_PATH=C:\boost_1_59_0   BOOST_LIBRARY_PATH=C:\boost_1_59_0\stage\lib   PYDOC=0   ARCH=sse3   BUILD_STATICLIB=yes   COMPILE_BOOST_TESTS=0   CPPPATH=E:\MATLAB\SharedSourceCode\Development\Mex\matlab\include,E:\MATLAB\SharedSourceCode\Development\Mex   LIBPATH=E:\MATLAB\SharedSourceCode\Development\Mex\matlab\bin   RPATH=E:\MATLAB\SharedSourceCode\Development\Mex\matlab\bin
Fatal Python error: Py_Initialize: unable to load the file system codec
ModuleNotFoundError: No module named 'encodings'

Current thread 0x00001c70 (most recent call first):
```

USE_PYTHON3_FOR_PYSGPP=0 and Py2.7 I was able to compile SGpp, but the build of the Python bindings failed:

```
$ ./schobepr_scons_build_mingw.bat

E:\MATLAB\SparseGridsLifecycle\BBSG\Examples\external\SGpp>scons -j 4   VERBOSE=0   SG_ALL=0   SG_BASE=1   SG_OPTIMIZATION=1   OPT=1   USE_EIGEN=1   USE_MATLAB=1   COMPILER=gnu   SG_PYTHON=1   USE_PYTHON3_FOR_PYSGPP=1   RUN_PYTHON_TESTS=0   BOOST_INCLUDE_PATH=C:\boost_1_59_0   BOOST_LIBRARY_PATH=C:\boost_1_59_0\stage\lib   PYDOC=0   ARCH=sse3   BUILD_STATICLIB=yes   COMPILE_BOOST_TESTS=0   CPPPATH=E:\MATLAB\SharedSourceCode\Development\Mex\matlab\include,E:\MATLAB\SharedSourceCode\Development\Mex   LIBPATH=E:\MATLAB\SharedSourceCode\Development\Mex\matlab\bin   RPATH=E:\MATLAB\SharedSourceCode\Development\Mex\matlab\bin
scons: Reading SConscript files ...
Info: Platform: Windows, YLPU060128, 7, 6.1.7601, AMD64, Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
Info: Using SCons 3.0.0 on Python 2.7.14
Info: SCons command line: 'C:\Users\pschober\AppData\Local\conda\conda\envs\Py27\Scripts\scons.py' -j 4 VERBOSE=0 SG_ALL=0 SG_BASE=1 SG_OPTIMIZATION=1 OPT=1 USE_EIGEN=1 USE_MATLAB=1 COMPILER=gnu SG_PYTHON=1 USE_PYTHON3_FOR_PYSGPP=1 RUN_PYTHON_TESTS=0 'BOOST_INCLUDE_PATH=C:\boost_1_59_0' 'BOOST_LIBRARY_PATH=C:\boost_1_59_0\stage\lib' PYDOC=0 ARCH=sse3 BUILD_STATICLIB=yes COMPILE_BOOST_TESTS=0 'CPPPATH=E:\MATLAB\SharedSourceCode\Development\Mex\matlab\include,E:\MATLAB\SharedSourceCode\Development\Mex' 'LIBPATH=E:\MATLAB\SharedSourceCode\Development\Mex\matlab\bin' 'RPATH=E:\MATLAB\SharedSourceCode\Development\Mex\matlab\bin'
Info: Available modules: SG_BASE, SG_COMBIGRID, SG_DATADRIVEN, SG_FINANCE, SG_MISC, SG_OPTIMIZATION, SG_PARALLEL, SG_PDE, SG_QUADRATURE, SG_SOLVER
Info: Available language support: SG_JAVA, SG_PYTHON

Checking programs and libraries:
Info: Using g++ 4.9.1
Checking for g++...(cached) C:\mingw64\bin\g++.EXE
Checking for gcc...(cached) C:\mingw64\bin\gcc.EXE
Checking for $SMARTLINK...(cached) C:\mingw64\bin\gcc.EXE
Checking for flag "-std=c++11"... (cached) yes
Checking for swig...(cached) C:\swigwin-3.0.8\swig.EXE
Info: Using SWIG 3.0.8
WindowsError: [Error 2] Das System kann die angegebene Datei nicht finden:
  File "E:\MATLAB\SparseGridsLifecycle\BBSG\Examples\external\SGpp\SConstruct", line 266:
    SGppConfigure.doConfigure(env, moduleFolders, languageSupport)
  File "E:\MATLAB\SparseGridsLifecycle\BBSG\Examples\external\SGpp\site_scons\SGppConfigure.py", line 100:
    checkPython(config)
  File "E:\MATLAB\SparseGridsLifecycle\BBSG\Examples\external\SGpp\site_scons\SGppConfigure.py", line 243:
    "import distutils.sysconfig; "
  File "C:\Users\pschober\AppData\Local\conda\conda\envs\Py27\lib\subprocess.py", line 212:
    process = Popen(stdout=PIPE, *popenargs, **kwargs)
  File "C:\Users\pschober\AppData\Local\conda\conda\envs\Py27\lib\subprocess.py", line 390:
    errread, errwrite)
  File "C:\Users\pschober\AppData\Local\conda\conda\envs\Py27\lib\subprocess.py", line 640:
    startupinfo)
```

I have not debugged anything yet, just fyi for now

**Comment by valentjn** [comment]
In GitLab by @valentjn on Dec 11, 2017, 09:51

* This issue is intended for a general discussion about when/how to support Python 3.x and what to migrate in SG++. Please open a new issue if you have any specific problems, sorry.
* SG++ officially only supports direct installations of Python (no distributions). We don't have any experience with Python distributions.
* That seems like a Python error which is independent of SG++. Try if just `scons` works.
* Google [1] says you should check `PYTHONPATH` and `PYTHONHOME`, they should point to Python 3.x.

[1] https://stackoverflow.com/a/5694730

**Comment by valentjn** [comment]
In GitLab by @schobepr on Dec 11, 2017, 09:55

1. Just running "scons" doesn't work either, gives the same error.
2. I tried the solution suggested [1]. Neither of the variables is set in my console. I can confirm I used the correct Python version.

I'll move this issue.

**Comment by valentjn** [comment]
In GitLab by @schobepr on Dec 11, 2017, 09:55

closed

**Comment by valentjn** [comment]
In GitLab by @schobepr on Dec 11, 2017, 09:56

reopened

**Comment by valentjn** [comment]
In GitLab by @valentjn on Dec 11, 2017, 09:56

Please don't just close the issues you were hijacking :wink:

**Comment by valentjn** [comment]
In GitLab by @valentjn on Jul 6, 2018, 11:22

There is now a branch `feature_python3`, which contains some commits that should enable it to use Python 3.x for both SCons and pysgpp. It should still be possible to use Python 2.x for one or even both of them. The Python scripts in `datadriven` etc. are not ported yet. (Tagging @freifrauvonbleifrei for interest.)

**Comment by valentjn** [comment]
In GitLab by @valentjn on Aug 22, 2018, 13:19

mentioned in merge request !73

**Comment by valentjn** [comment]
In GitLab by @freifrauvonbleifrei on Aug 22, 2018, 13:20

mentioned in merge request !74

**Comment by valentjn** [comment]
Finally, this has been fixed by #90/#91 with follow-up fixes in #94.

For the record: Python 2.x isn't supported anymore for the SWIG Python interface. It's probably still possible to run the SCons scripts with Python 2.x, but I wouldn't rely on that.
